### PR TITLE
bump*: tweak autobump message

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

We've had multiple issues filed because BrewTestBot isn't bumping
formulae, and they're not able to `brew bump` it.

Let's adjust the messaging to encourage them to open a PR manually
instead.

We should probably also make BrewTestBot ignore open PRs that aren't
version-bump PRs, but I'd prefer to punt on that for now.

### CODE REVIEW

**Comment by Bo98** [comment]
Feels really weird we're going through all this effort to direct people to manual PRs that often can be lower quality. 🤷‍♂️

**Review by Bo98 — APPROVED** [review]

**Comment by carlocab** [comment]
> Feels really weird we're going through all this effort to direct people to manual PRs that often can be lower quality. 🤷‍♂️

Agreed. But even removing the autobump restriction doesn't do much because we've also gotten rid of `--force`, so we'd need to put `--force` back too.

I'm game to do that, but I'm not sure others would be.

Did we have lots of contributors using `--force` when they shouldn't be? I don't really recall any.

**Comment by Bo98** [comment]
I've never really been convinced, even when we chatted about it at the AGM, so I'm perhaps not the best to ask.

I don't really have any strong feelings other than the feeling of sympathy for end-users that are frustrated by the bot not being 100% perfect but blocking anyone from correcting it.

**Comment by carlocab** [comment]
I don't even really recall the `--force` conversation at the AGM.

Thoughts, @Homebrew/brew, @Homebrew/core, @Homebrew/cask?

**Comment by Bo98** [comment]
Was more referring to the whole blocking people from doing `brew bump` thing.

**Comment by bevanjkay** [comment]
From a personal perspective, the removal of the `--force` flag has meant some additional work to open PRs when opening them manually is required. In the past I could just pass `--force`, where now have to find out the new checksums, edit the files manually, and then generate commits etc...

From memory, the primary concern was around contributors run `brew bump` and opening PRs for absolutely everything. We have some steps in place for rate-limiting now (setting a max open PR limit), so perhaps this isn't that great of a concern anymore.

**Comment by carlocab** [comment]
Yea having to create PRs manually is pretty annoying.

I feel like we should at least allow `--force` for `bump-*-pr` but leave it out for `bump`.

It should be possible to make detection of version bump PRs better too, but I'd need to find the time for it.

**Comment by Bo98** [comment]
There's two separate issues here:

* One is the bot failing to occasionally bump (livecheck failure etc.)
* The other is the bot getting blocked by existing PRs for unrelated things

The latter is partially by design. Duplicate checking is a fundamental part of how BrewTestBot doesn't issue the same version bump over and over again. We could however perhaps make the "open duplicate PR" check more like the "closed duplicate PR" check for the bot.

---

> From memory, the primary concern was around contributors run `brew bump` and opening PRs for absolutely everything.

In the end the volume of PRs is largely identical. All that's changed is the PR author.

We can also never really guarantee the bot will be 100% perfect. Rate limiting etc can mean the bot can suddenly not pick up certain versions without us knowing. Livechecks can break etc.

> I feel like we should at least allow `--force` for `bump-*-pr` but leave it out for `bump`. Or, at least, allow `--force` for maintainers so we can `bump-*-pr --force` as soon as someone complains about a missing version bump from the autobump list. (Or make BrewTestBot do it while ignoring open PRs.)

To clarify here: `--force` was removed before the autobump check was added. So `--force` to override the autobump restriction never existed - it was only for the duplicate PR check.

But yes, I do see the angle on this anyhow. I reckon we can address this without needing to resort a `--force` flag, but if we did anyway we can also detect abuse by adding a "This PR was force created" to the PR description.

**Comment by carlocab** [comment]
> There's two separate issues here:
> 
> * One is the bot failing to occasionally bump (livecheck failure etc.)
> * The other is the bot getting blocked by existing PRs for unrelated things

In my experience it's really the latter that is driving the issues/complaints that we get, but I can see how the former would be an issue for the way we handle autobumps too.

> but if we did add it anyway we can also detect abuse by adding a "This PR was force created" to the PR description.

Yea, I was definitely thinking this.

**Comment by Bo98** [comment]
> In my experience it's really the latter that is driving the issues/complaints that we get, but I can see how the former would be an issue for the way we handle autobumps too.

There's been multiple reports in Homebrew/cask on IntelliJ casks not updating.

I guess the former problem is more common for casks, and the latter problem is more common for formulae.

**Comment by Bo98** [comment]
> I guess the former problem is more common for casks, and the latter problem is more common for formulae.

Or well perhaps it does happen a lot to formulae too. The latest run shows _23_ formulae failed version checking: https://github.com/Homebrew/homebrew-core/actions/runs/10626272934/job/29457555049

**Comment by SMillerDev** [comment]
> In the end the volume of PRs is largely identical. All that's changed is the PR author.

Which is actually a big change because we know we can't expect testbot to fix a PR, while some random drive by might take a lot of handholding to never get merged in the end.

As for the duplication check, I think that should only check for exact version matches, because currently it's blocking a major release if you have a typo fix open.

**Comment by Bo98** [comment]
> > In the end the volume of PRs is largely identical. All that's changed is the PR author.
> 
> Which is actually a big change because we know we can't expect testbot to fix a PR, while some random drive by might take a lot of handholding to never get merged in the end.

That's a useful perspective, thanks!

Maybe a force override flag will end up being best then.

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]: Not your fault but: this is way too much text to output here. My experience is the more text you output in these sort of cases, the less likely people are to read any of it. I'd rather we fixed the underlying issues here rather than improved our messaging around them.

**Comment by MikeMcQuaid** [comment]
> We should probably also make BrewTestBot ignore open PRs that aren't
> version-bump PRs, but I'd prefer to punt on that for now.

@carlocab This seems like the main fix to be made here so it'd be good to do it instead.

> Or well perhaps it does happen a lot to formulae too. The latest run shows 23 formulae failed version checking: [Homebrew/homebrew-core/actions/runs/10626272934/job/29457555049](https://github.com/Homebrew/homebrew-core/actions/runs/10626272934/job/29457555049)

@Bo98 What does "failed version checking" mean here, how can I see what these 23 are and how/where/are we notified?

This feels like the other problem to be addressed: we have increasing number of scheduled jobs which, pretty much with the sole exception of formulae.brew.sh which creates an issue, go completely unnoticed when they are broken/not working as expected. It seems like _all_ scheduled issues should _always_ create an issue if they are failing and, if the "23 formulae failed" above: that should have opened an issue and made that job have 🔴 failed CI status.

> * One is the bot failing to occasionally bump (livecheck failure etc.)

These should be either fixed or removed from the autobump list if they are not being reliably bumped. Again, I think re-adding `--force` or adding more messaging papers over an actual problem here that needs fixing.

**Comment by MikeMcQuaid** [comment]
> > We should probably also make BrewTestBot ignore open PRs that aren't
> > version-bump PRs, but I'd prefer to punt on that for now.
> 
> @carlocab This seems like the main fix to be made here so it'd be good to do it instead.

Done this in https://github.com/Homebrew/brew/pull/18206. We're only going to fail if we fail on an exact match instead.

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]: @carlocab Yeh, fair, just think https://github.com/Homebrew/brew/pull/18206 may be a better fit here. Let me know what you think.

Comment (carlocab) [reply Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]: Thanks; will have a look later today!

**Comment by carlocab** [comment]
Closing in favour of #18206.

**Comment by chenrui333** [comment]
> > In the end the volume of PRs is largely identical. All that's changed is the PR author.
> 
> Which is actually a big change because we know we can't expect testbot to fix a PR, while some random drive by might take a lot of handholding to never get merged in the end.

IIRC, another thing is the bottle cache, it would be really tricky for non-maintainer to open a PR which has lots of dependants to test.

**Pull request closed without merging**
